test/retry/main.ml

```diff
@@ -0,0 +1,158 @@
+                 Lwt.return_error (`Fatal ()))
+        in
+        let* n_retry_errors = Lwt_stream.nget retries_before_fatal strm >|= List.length in
+        assert (n_retry_errors = retries_before_fatal);
+        let* fatal_error = Lwt_stream.next strm in
+        assert (fatal_error = Error (`Fatal (), retries_before_fatal + 1));
+        (* ensure the post condition of an empty stream *)
+        Lwt_stream.is_empty strm);
+
+    test "retries before success" (fun () ->
+        let retries_before_fatal = 3 in
+        let i = ref 0 in
+        let strm = Retry.on_error (fun () ->
+            if !i < retries_before_fatal then (
+              incr i;
+              Lwt.return_error (`Retry ())
+            ) else
+              Lwt.return_ok ()
+          )
+        in
+        let* n_retry_errors = Lwt_stream.nget retries_before_fatal strm >|= List.length in
+        assert (n_retry_errors = retries_before_fatal);
+        let* success = Lwt_stream.next strm in
+        assert (success = Ok ());
+        (* ensure the post condition of an empty stream *)
+        Lwt_stream.is_empty strm);
+
+    test "[n_times 0] runs one attempt" (fun () ->
+        let operation () = Lwt.return_error (`Retry ()) in
+        let+ attempt = Retry.(operation |> on_error |> n_times 0) in
+        attempt = Error (`Retry (), 1));
+
+    test "n_times gives up on a fatal error" (fun () ->
+        let i = ref 0 in
+        let operation () =
+          if !i < 3 then (
+            incr i;
+            Lwt.return_error (`Retry ())
+          ) else
+            Lwt.return_error (`Fatal ())
+        in
+        let+ fatal_error = Retry.(operation |> on_error |> n_times 5) in
+        fatal_error = Error (`Fatal (), 4));
+
+    test "n_times gives a retry error when exhausted"  (fun () ->
+        let retries = 5 in
+        let operation () = Lwt.return_error (`Retry ()) in
+        let+ result = Retry.(operation |> on_error |> n_times retries) in
+        result = Error (`Retry (), retries + 1));
+
+    test "n_times is ok on success" (fun () ->
+        let i = ref 0 in
+        let operation () =
+          if !i < 3 then (
+            incr i;
+            Lwt.return_error (`Retry ())
+          ) else
+            Lwt.return_ok ()
+        in
+        let+ success = Retry.(operation |> on_error |> n_times 5) in
+        success = Ok ());
+
+    test_direct "n_times on negative raises Invalid_argument"  (fun () ->
+        let invalid_negative_retries = -5 in
+        let operation () = Lwt.return_error (`Retry ()) in
+        let attempts = Retry.(operation |> on_error) in
+        try
+          let _ = Retry.(attempts |> n_times invalid_negative_retries) in
+          false (* We failed to raise the invalid argument exception *)
+        with
+          Invalid_argument _ -> true);
+
+    (* test that the sleeps actually throttle computations as desired *)
+    test "with_sleep really does sleep"  (fun ()  ->
+      let duration _ = 0.1 in
```

Comment:
```suggestion
      let duration _ = 0.01 in
```

This is the bottleneck for the test suite and it doesn't make sense to go for tenth of a seconds order of magniture.